### PR TITLE
Quartz scheduler approach for deleting Inactive consumers in Strimzi Bridge

### PR DESCRIPTION
**Pull request by ShubhamRwt**
This PR adds the mechanism to automatically shut down the consumer if it is running and has passed out the given timeout.

The timeout is inactive by default.

The timeout should be given in second in the application.properties file like this

http.timeoutSeconds = 3000

### CODE REVIEW

**Comment by ShubhamRwt** [comment]
I have added the changes @ppatierno

File: pom.xml [hunk 1]
```diff
@@ -145,6 +145,15 @@
 			<groupId>io.quarkus</groupId>
 			<artifactId>quarkus-resteasy-reactive-jackson</artifactId>
 		</dependency>
+		<dependency>
+			<groupId>io.quarkus</groupId>
+			<artifactId>quarkus-scheduler</artifactId>
```
Comment (ppatierno) [inline pom.xml]: genuine question: when not using the `@Schedule` annotation, there are still classes we are using that need this dependency? Isn't the following quartz dependency enough?

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 2]
```diff
@@ -60,6 +69,9 @@ public class RestBridge {
     @Inject
     Logger log;
 
+    @Inject
+    org.quartz.Scheduler quartz;
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: why the FQN and not just `Scheduler`?

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 3]
```diff
@@ -87,7 +99,12 @@ public void init() {
         adminClientEndpoint.open();
 
         if (this.httpConfig.timeoutSeconds() > -1) {
-            startInactiveConsumerDeletionTimer(this.httpConfig.timeoutSeconds());
+            try {
+                jobScheduler(this.httpConfig.timeoutSeconds());
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: let's call it as before `startInactiveConsumerDeletionTimer`

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 4]
```diff
@@ -87,7 +99,12 @@ public void init() {
...
+                System.out.println(this.httpConfig.timeoutSeconds());
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: can we remove this one I guess

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 5]
```diff
@@ -331,7 +348,7 @@ public CompletionStage<Response> ready() {
      * Retrieves the sink endpoint based on the consumer group and the instance name
      *
      * @param groupId consumer group
-     * @param name instance name
+     * @param name    instance name
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: why this change?

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 6]
```diff
@@ -357,7 +374,7 @@ private RestSinkBridgeEndpoint<byte[], byte[]> getRestSinkBridgeEndpoint(String
      * It creates and returns a new source endpoint, if the HTTP request comes from a new HTTP connection.
      *
      * @param routingContext RoutingContext instance used for getting the HTTP connection and attaching a close handler to it
-     * @param contentType Content-Type header from the HTTP request to be mapped to the corresponding embedded content type
+     * @param contentType    Content-Type header from the HTTP request to be mapped to the corresponding embedded content type
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: why this change?

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 7]
```diff
@@ -472,7 +489,7 @@ private RestSinkBridgeEndpoint<byte[], byte[]> doCreateConsumer(JsonNode jsonBod
      * Return the sink endpoint which has to be closed together with its underneath Kafka Consumer
      *
      * @param groupId consumer group
-     * @param name name of the sink endpoint instance to delete within the consumer group
+     * @param name    name of the sink endpoint instance to delete within the consumer group
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: why this change?

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 8]
```diff
@@ -549,7 +566,47 @@ private boolean isReady() {
         return this.isReady;
     }
 
-    private void startInactiveConsumerDeletionTimer(Long timeout) {
-        // TODO: to be used a Quarkus timer for scheduling the task
+    private void jobScheduler(long timeout) throws SchedulerException {
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: as said above let's name it `startInactiveConsumerDeletionTimer` as before.

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 9]
```diff
@@ -549,7 +566,47 @@ private boolean isReady() {
...
+                    deleteSinkEndpoint.close();
+                    httpBridgeContext.getHttpSinkEndpoints().remove(item.getKey());
+                    log.infof("Consumer %s deleted after inactivity timeout.", item.getKey());
+                    timestampMap.remove(item.getKey());
+                }
+            }
+        }
+    }
+
+    public static class MyJob implements Job {
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: I guess we need a more explicit name, not just `MyJob`, or?
Even adding some Javadoc would help.

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 10]
```diff
@@ -549,7 +566,47 @@ private boolean isReady() {
...
+                                .withIntervalInMilliseconds(timeInMs / 2)
+                                .repeatForever())
+                .build();
+        quartz.scheduleJob(job, trigger);
+    }
+
+    void performTask() {
+        long timeoutInMs = httpConfig.timeoutSeconds() * 1000L;
+
+        log.infof("Looking for stale consumers in %s entries", timestampMap.size());
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: I put it as `infof` just for testing, we should use debug instead.

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 11]
```diff
@@ -549,7 +566,47 @@ private boolean isReady() {
...
+        Iterator<Map.Entry<ConsumerInstanceId, Long>> it = timestampMap.entrySet().iterator();
+        while (it.hasNext()) {
+            Map.Entry<ConsumerInstanceId, Long> item = it.next();
+            if (item.getValue() + timeoutInMs < System.currentTimeMillis()) {
+                RestSinkBridgeEndpoint<byte[], byte[]> deleteSinkEndpoint = httpBridgeContext.getHttpSinkEndpoints().get(item.getKey());
+                if (deleteSinkEndpoint != null) {
+                    deleteSinkEndpoint.close();
+                    httpBridgeContext.getHttpSinkEndpoints().remove(item.getKey());
+                    log.infof("Consumer %s deleted after inactivity timeout.", item.getKey());
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: ditto as above

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 12]
```diff
@@ -549,7 +566,47 @@ private boolean isReady() {
...
+                .startNow()
+                .withSchedule(
+                        SimpleScheduleBuilder.simpleSchedule()
+                                .withIntervalInMilliseconds(timeInMs / 2)
+                                .repeatForever())
+                .build();
+        quartz.scheduleJob(job, trigger);
+    }
+
+    void performTask() {
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: some Javadoc

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java (hunk 5 above)
Comment (ShubhamRwt) [reply src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: Ahh its due to my new Mac . I will fix these

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java (hunk 6 above)
Comment (ShubhamRwt) [reply src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: I guess we can keep these changes since they align the descriptions of the parameter? We follow this pattern in the operator also.

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java (hunk 2 above)
Comment (ShubhamRwt) [reply src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: Sorry, I didn't got this one?

Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: I mean why not just ....

```java
@Inject
Scheduler quartz;
```

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java (hunk 6 above)
Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: I would avoid this for now. Let's keep less changes and focus on what is the PR.

File: pom.xml [hunk 13]
```diff
@@ -145,6 +145,11 @@
 			<groupId>io.quarkus</groupId>
 			<artifactId>quarkus-resteasy-reactive-jackson</artifactId>
 		</dependency>
+		<dependency>
+			<groupId>io.quarkus</groupId>
+			<artifactId>quarkus-quartz-deployment</artifactId>
+		</dependency>
+
```
Comment (ppatierno) [inline pom.xml]: why this extra line?
Also, we are using Quartz now, from the doc it seems that we need the `quarkus-quartz` dependency (see here https://quarkus.io/guides/quartz). What is this `quarkus-quartz-deployment`?

**Comment by ShubhamRwt** [comment]
@ppatierno Should I mention in the `application.properties` that for setting the timeout we have to use this --> `quarkus.scheduler.enabled=false `

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java (hunk 6 above)
Comment (ShubhamRwt) [reply src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: Okay, let me revert it

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java (hunk 2 above)
Comment (ShubhamRwt) [reply src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: Yeah done, Sorry I was confused by FQN

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 14]
```diff
@@ -549,7 +566,53 @@ private boolean isReady() {
         return this.isReady;
     }
 
-    private void startInactiveConsumerDeletionTimer(Long timeout) {
-        // TODO: to be used a Quarkus timer for scheduling the task
+    private void startInactiveConsumerDeletionTimer(long timeout) throws SchedulerException {
+        long timeInMs = timeout * 1000L;
+        JobDetail job = JobBuilder.newJob(InactiveConsumerDeletionJob.class).build();
+        Trigger trigger = TriggerBuilder.newTrigger()
+                .withIdentity("myTrigger", "myGroup")
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: can we use better names for "myTrigger" and "myGroup"?

Comment (ShubhamRwt) [reply src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: "triggerInactiveConsumerDeletion" for `myTrigger` and `staleConsumers` for `myGroup` ?

Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: fine with me :-)

File: config/application.properties [hunk 15]
```diff
@@ -38,4 +38,8 @@ quarkus.application.name=strimzi-kafka-bridge
 #using OTEL_SERVICE_NAME as usual or a preformatted service name based on the bridge.id
 quarkus.opentelemetry.tracer.resource-attributes=service.name=${OTEL_SERVICE_NAME:strimzi-kafka-bridge-${bridge.id}}
 #if the OTEL_EXPORTER_OTLP_ENDPOINT, it's using http://localhost:4317 by default
-quarkus.opentelemetry.tracer.exporter.otlp.endpoint=${OTEL_EXPORTER_OTLP_ENDPOINT}
\ No newline at end of file
+quarkus.opentelemetry.tracer.exporter.otlp.endpoint=${OTEL_EXPORTER_OTLP_ENDPOINT}
+
+#Inactive consumer timeout configuration
+#This property is added so that you can configure the http.timeoutSeconds the usual way.
```
Comment (ppatierno) [inline config/application.properties]: Tbh I think we should explain differently. Not related to the timeout but saying that it's needed to allow scheduling a periodic task programmatically.

**Comment by ppatierno** [comment]
@ShubhamRwt I had to rebase and now other conflicts. Anyway it sounds really strange to see 32 commits here instead of just the one you made. I think the PR should be cleaned from this point of view.

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 16]
```diff
@@ -548,7 +565,53 @@ private boolean isReady() {
...
+                                .withIntervalInMilliseconds(timeInMs / 2)
+                                .repeatForever())
+                .build();
+        quartz.scheduleJob(job, trigger);
+    }
+
+    /**
+     * Looks up for the stale consumers and remove them if they have passed the timeout
+     */
+    void performTask() {
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: it could be something better like `deleteInactiveConsumers`?

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 17]
```diff
@@ -548,7 +565,53 @@ private boolean isReady() {
...
+                .withSchedule(
+                        SimpleScheduleBuilder.simpleSchedule()
+                                .withIntervalInMilliseconds(timeInMs / 2)
+                                .repeatForever())
+                .build();
+        quartz.scheduleJob(job, trigger);
+    }
+
+    /**
+     * Looks up for the stale consumers and remove them if they have passed the timeout
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: let's use `inactive` instead of `stale` as we do in other places. Wdyt?

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 18]
```diff
@@ -548,7 +565,53 @@ private boolean isReady() {
...
+     */
+    void performTask() {
+        long timeoutInMs = httpConfig.timeoutSeconds() * 1000L;
+
+        log.debugf("Looking for stale consumers in %s entries", timestampMap.size());
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: `inactive` again

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 19]
```diff
@@ -548,7 +565,53 @@ private boolean isReady() {
...
+                    log.debugf("Consumer %s deleted after inactivity timeout.", item.getKey());
+                    timestampMap.remove(item.getKey());
+                }
+            }
+        }
+    }
+
+    /**
+     * Job class which contains the task to be run periodically by the scheduler
+     */
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java:606]: This should actually explain what is the Job about ... so deleting the inactive consumers.

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 20]
```diff
@@ -548,7 +565,53 @@ private boolean isReady() {
         return this.isReady;
     }
 
-    private void startInactiveConsumerDeletionTimer(Long timeout) {
-        // TODO: to be used a Quarkus timer for scheduling the task
+    private void startInactiveConsumerDeletionTimer(long timeout) throws SchedulerException {
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: let's use the plural and change to `scheduleInactiveConsumersDeletionJob`, wdyt?

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 21]
```diff
@@ -548,7 +565,53 @@ private boolean isReady() {
...
+        long timeInMs = timeout * 1000L;
+        JobDetail job = JobBuilder.newJob(InactiveConsumerDeletionJob.class).build();
+        Trigger trigger = TriggerBuilder.newTrigger()
+                .withIdentity("inactiveConsumerDeletion", "staleConsumers")
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: use plural on "Consumers"

File: src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java [hunk 22]
```diff
@@ -548,7 +565,53 @@ private boolean isReady() {
...
+        log.debugf("Looking for stale consumers in %s entries", timestampMap.size());
+        Iterator<Map.Entry<ConsumerInstanceId, Long>> it = timestampMap.entrySet().iterator();
+        while (it.hasNext()) {
+            Map.Entry<ConsumerInstanceId, Long> item = it.next();
+            if (item.getValue() + timeoutInMs < System.currentTimeMillis()) {
+                RestSinkBridgeEndpoint<byte[], byte[]> deleteSinkEndpoint = httpBridgeContext.getHttpSinkEndpoints().get(item.getKey());
+                if (deleteSinkEndpoint != null) {
+                    deleteSinkEndpoint.close();
+                    httpBridgeContext.getHttpSinkEndpoints().remove(item.getKey());
+                    log.debugf("Consumer %s deleted after inactivity timeout.", item.getKey());
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java]: use `warnf` here, the user should always see when consumers are deleted.

**Review by ppatierno — APPROVED** [review]
LGTM. Thanks!
